File: implementations/flutter/design_patterns_flutter/lib/patterns/structural/bridge/abstraction.dart
import 'domain/report_data.dart';
import 'implementor.dart';

abstract class Report {
  final Exporter exporter;
  Report(this.exporter);

  Future<ReportData> build();

  Future<ExportResult> generateReport() async {
    final data = await build();
    return exporter.export(data);
  }
}

File: implementations/react-native/design-patterns-rn/src/patterns/structural/bridge/abstraction.ts
import type { Exporter } from './implementor';
import type { ReportData } from './interfaces/report-data';

export abstract class Report {
  protected exporter: Exporter;
  constructor(exporter: Exporter) { this.exporter = exporter; }

  abstract build(): Promise<ReportData>;

  async generateReport() {
    const data = await this.build();
    return this.exporter.export(data);
  }
}
